android/uhabits-core/src/main/java/org/isoron/uhabits/core/io/HabitsCSVImporter.java

```diff
@@ -0,0 +1,160 @@
+
+                int idx = 0;
+                String position = line[idx++];
+                String name = line[idx++];
+                String question = hasQuestion ? line[idx++] : null;
+                String desc = line[idx++];
+                int reps = Integer.parseInt(line[idx++]);
+                int intv = Integer.parseInt(line[idx++]);
+                String colr = line[idx++];
+
+                Habit habit = findExisting(name);
+                if (habit == null) {
+                    habit = modelFactory.buildHabit();
+                    habit.setName(name);
+                    if (hasQuestion) {
+                        habit.setDescription(desc);
+                        habit.setQuestion(question);
+                    } else {
+                        habit.setDescription("");
+                        habit.setQuestion(desc);
+                    }
+                    habit.setFrequency(new Frequency(reps, intv));
+                    habit.setColor(findColor(colr));
+                    habitList.add(habit);
+                }
+
+                parseCheckmarks(findCheckmarks(position, zipFile), habit);
+            }
+        } finally {
+            try {
+                zipFile.close();
+                habitsCsv.close();
+            } catch (IOException e) {
+                // ignore
+            }
+        }
+    }
+
+    Timestamp parse(String ts) throws ParseException {
+        Date date = DateFormats.getCSVDateFormat().parse(ts);
+        Timestamp timestamp = new Timestamp(date.getTime());
+        return timestamp;
+    }
+
+    void parseCheckmarks(CSVReader checkmarks, Habit habit) {
+        if (checkmarks == null) return;
+        try {
+            for (String line[] : checkmarks) {
+                try {
+                    Timestamp ts = parse(line[0]);
+                    int value = Integer.parseInt(line[1]);
+                    // only add positive changes ;)
+                    if (value > Checkmark.UNCHECKED) {
+                        habit.getRepetitions().toggle(ts, value);
+                    }
+                } catch (Exception e) {
+                    // skip
+                }
+            }
+        } finally {
+            try {
+                checkmarks.close();
+            } catch (IOException e) {
+                // ignore
+            }
+        }
+    }
+
+    CSVReader findCheckmarks(String position, ZipFile zipFile) throws IOException {
+        Enumeration<? extends ZipEntry> entries = zipFile.entries();
+        while (entries.hasMoreElements()) {
+            ZipEntry entry = entries.nextElement();
+            // look for 00x */Checkmarks.csv
+            if (!entry.isDirectory() && entry.getName().startsWith(position) && entry.getName().endsWith("/Checkmarks.csv")) {
+                return new CSVReader(new InputStreamReader(zipFile.getInputStream(entry)));
+            }
+        }
+        return null;
+    }
+
+    private Habit findExisting(String name) {
```

Comment:
I don't really know how this plays out performance wise to loop over the habitList over and over again?